craft_parts/plugins/rust_plugin.py

```diff
@@ -111,32 +128,109 @@ def get_build_snaps(self) -> Set[str]:
     @override
     def get_build_packages(self) -> Set[str]:
         """Return a set of required packages to install in the build environment."""
-        return {"gcc", "git"}
+        return {"curl", "gcc", "git", "pkg-config", "findutils"}
+
+    def _check_system_rust(self) -> bool:
+        """Check if Rust is installed on the system."""
+        try:
+            rust_version = subprocess.check_output(["rustc", "--version"])
+            cargo_version = subprocess.check_output(["cargo", "--version"])
+            return "rustc" in rust_version.decode(
+                "utf-8"
+            ) and "cargo" in cargo_version.decode("utf-8")
+        except (subprocess.CalledProcessError, FileNotFoundError):
+            return False
+
+    def _check_rustup(self) -> bool:
+        try:
+            rustup_version = subprocess.check_output(["rustup", "--version"])
+            return "rustup" in rustup_version.decode("utf-8")
+        except (subprocess.CalledProcessError, FileNotFoundError):
+            return False
+
+    def _get_setup_rustup(self, channel: str) -> List[str]:
+        return [
+            f"""\
+curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | \
+sh -s -- -y --no-modify-path --profile=minimal --default-toolchain {channel}
+"""
+        ]
 
     @override
     def get_build_environment(self) -> Dict[str, str]:
         """Return a dictionary with the environment to use in the build step."""
-        return {"PATH": "${HOME}/.cargo/bin:${PATH}"}
+        return {
+            "PATH": "${HOME}/.cargo/bin:${PATH}",
+        }
+
+    @override
+    def get_pull_commands(self) -> List[str]:
+        """Return a list of commands to run during the pull step."""
+        options = cast(RustPluginProperties, self._options)
+        if not options.rust_channel and self._check_system_rust():
```

Comment:
also check if self._part.dependencies has rust-deps

Reply:
`self._part_info` is available to plugins (instead of `self_part`) and it currently doesn't provide dependencies. The information is currently passed only to environment validators. @liushuyu may have an alternative solution for this, otherwise we can add dependency information to `PartInfo`.

Reply:
I have pushed an alternative solution yesterday:

https://github.com/canonical/craft-parts/pull/530/files#diff-f7845e9499f862b5d9c9c6140de217cea696e6c67f6dfb03cfe1bdf421861486R88-R99

Reply:
It seems that current solution detects if both the channel and a rust-deps part are available, however if only the rust-deps part is used, rustup is still invoked? `self._check_system_rust()` would fail at this point because the rust-deps part is not staged yet.

Reply:
> It seems that current solution detects if both the channel and a rust-deps part are available, however if only the rust-deps part is used, rustup is still invoked, `self._check_system_rust()` would fail at this point because the rust-deps part is not staged yet.

Addressed.
Now it will automatically set `"none"` in `validate_environment` if the user has `rust-deps` defined.